core/lib/prover_interface/src/outputs.rs

```diff
@@ -120,14 +123,57 @@ impl fmt::Debug for L1BatchTeeProofForL1 {
     }
 }
 
+#[derive(Copy, Clone)]
+pub enum FinalProofKeyBySubsystem {
```

Comment:
hmm why do you expose `protocol_version` and `batch_number`? You do multiple `match`es - note the `encode_key` method below - you first `protocol_version()`  there that results in a `match` **inside** `FinalProofKeyBySubsystem` but then also do a `match` over the `key` instance. Thus, the logic that build the actual file name from key is scattered between `FinalProofKeyBySubsystem` and `L1BatchProofForL1 ` and to follow it, one needs to look into both struct.

I suggest having `FinalProofKeyBySubsystem` as data struct instead (no logic) and have clear logic in `encode_key` method.

As a side note - instead we could put this logic to the `Key` type itself, but this will require refacatoring of all the `impl StoredObject for` blocks.

Reply:
Another side note - you could have a `trait` instead. This would make Prover and Core less coupled - although IMO both are accetptable